lua/gp/logger.lua

```diff
@@ -0,0 +1,71 @@
+local M = {}
+
+local file = "/dev/null"
+
+M._log_history = {}
+
+---@param path string # path to log file
+M.set_log_file = function(path)
+	M.debug("New neovim instance started, setting log file to " .. path)
+	local dir = vim.fn.fnamemodify(path, ":h")
+	if vim.fn.isdirectory(dir) == 0 then
+		vim.fn.mkdir(dir, "p")
+	end
+	file = path
+
+	local log_file = io.open(file, "a")
+	if log_file then
+		for _, line in ipairs(M._log_history) do
+			log_file:write(line .. "\n")
+		end
+		log_file:close()
+	end
+end
+
+---@param msg string # message to log
+---@param level integer # log level
+---@param slevel string # log level as string
+local log = function(msg, level, slevel)
+	local raw = string.format("[%s] %s: %s", os.date("%Y-%m-%d %H:%M:%S"), slevel, msg)
+
+	M._log_history[#M._log_history + 1] = raw
+	if #M._log_history > 100 then
+		table.remove(M._log_history, 1)
+	end
+
+	local log_file = io.open(file, "a")
+	if log_file then
+		log_file:write(raw .. "\n")
+		log_file:close()
+	end
+
+	if level == vim.log.levels.DEBUG then
```

Comment:
note that there is TRACE level too so we might be better off using >=.
Alternatively there is should_log from upstream https://github.com/neovim/neovim/blob/9793a97778dbb8b023272553dba6da87ae210ad3/runtime/lua/vim/lsp/log.lua#L173

Reply:
`<=` since we don't want to notify users about trace/debug
```
{
  DEBUG = 1,
  ERROR = 4,
  INFO = 2,
  OFF = 5,
  TRACE = 0,
  WARN = 3
}
```